README.md
[![Latest Version](https://img.shields.io/github/release/spatie/statamic-responsive-images.svg?style=flat-square)](https://github.com/spatie/statamic-responsive-images/releases)
![Statamic 3.0+](https://img.shields.io/badge/Statamic-3.0+-FF269E?style=flat-square&link=https://statamic.com)

# Algolia Places

> Algolia Places fieldtype for Statamic 3.

## Support us

[<img src="https://github-ads.s3.eu-central-1.amazonaws.com/statamic-algolia-places.jpg?t=1" width="419px" />](https://spatie.be/github-ad-click/statamic-algolia-places)

We invest a lot of resources into creating [best in class open source packages](https://spatie.be/open-source). You can support us by [buying one of our paid products](https://spatie.be/open-source/support-us).

We highly appreciate you sending us a postcard from your hometown, mentioning which of our package(s) you are using. You'll find our address on [our contact page](https://spatie.be/about-us). We publish all received postcards on [our virtual postcard wall](https://spatie.be/open-source/postcards).

## Installation

Require it using Composer.

```
composer require spatie/statamic-algolia-places
```

## Templating

The location field will contain the following data:

```yaml
location:
  name: 'Kruikstraat 22'
  administrative: Vlaanderen
  county: Antwerpen
  city: Antwerpen
  country: België
  countryCode: be
  type: address
  latlng:
    lat: 51.2088
    lng: 4.42386
  postcode: '2018'
  postcodes:
    - '2018'
  query: 'Kruikstraat 22'
  value: 'Kruikstraat 22, Antwerpen, Vlaanderen, België'
```

## Configuration

The fieldtype has all the [Algolia Places options](https://community.algolia.com/places/documentation.html#options)

![Config](https://raw.githubusercontent.com/spatie/statamic-algolia-places/master/docs/config.png)

### Changelog

Please see [CHANGELOG](https://github.com/spatie/statamic-algolia-places/blob/master/CHANGELOG.md) for more information what has changed recently.

## Contributing

Please see [CONTRIBUTING](https://github.com/spatie/statamic-algolia-places/blob/master/CONTRIBUTING.md) for details.

## Security

If you discover any security related issues, please email [freek@spatie.be](mailto:freek@spatie.be) instead of using the issue tracker.

## Credits

- [Rias Van der Veken](https://github.com/riasvdv)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](https://github.com/spatie/statamic-algolia-places/blob/master/LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `statamic-algolia-places` will be documented in this file

## 1.1.0 - 2021-01-29

## What's new
- Allow PHP 8

## 1.0.2 - 2020-11-03

## What's new
- The fieldtype now uses the selected site's locale by default

## 1.0.1 - 2020-10-23

## What's fixed
- Rename fieldtype to `places`

## 1.0.0 - 2020-10-23

## What's new
- Release!


src/Fieldtypes/Location.php
<?php

namespace Spatie\StatamicAlgoliaPlaces\Fieldtypes;

use Statamic\Fields\Fieldtype;

class Location extends Fieldtype
{
    protected $icon = 'pin';

    protected function configFieldItems(): array
    {
        return [
            'placeholder' => [
                'type' => 'text',
                'display' => __('Placeholder'),
                'default' => __('Start typing your address'),
            ],
            'searchType' => [
                'type' => 'select',
                'display' => __('Type'),
                'instructions' => __('Restrict the search results to a specific type'),
                'options' => [
                    'all' => 'All',
                    'city' => 'City',
                    'country' => 'Country',
                    'address' => 'Address',
                    'busStop' => 'Bus Stop',
                    'trainStation' => 'Train Station',
                    'townhall' => 'Townhall',
                    'airport' => 'Airport',
                ],
                'default' => 'all',
            ],
            'postcodeSearch' => [
                'type' => 'toggle',
                'display' => __('Postcode Search'),
                'instructions' => __('Only search in the postcode field'),
            ],
            'language' => [
                'type' => 'text',
                'display' => __('Language'),
                'instructions' => __('Change the default language of the results. You can pass two letters language codes (ISO 639-1).'),
                'default' => app()->getLocale(),
                'width' => 50,
            ],
            'countries' => [
                'type' => 'taggable',
                'display' => __('Countries'),
                'instructions' => __('Change the countries to search in. Two letters country codes (ISO 639-1).'),
                'width' => 50,
            ],
            'locationSearch' => [
                'type' => 'select',
                'options' => [
                    'no' => __('No'),
                    'lat_lng' => __('Around LatLng'),
                    'lat_lng_ip' => __('Around LatLng via IP'),
                    'bounding_box' => __('Inside a bounding box'),
                    'polygon' => __('Inside a polygon'),
                ],
                'default' => 'lat_lng_ip',
                'display' => __('Search around a location?'),
            ],
            'aroundLatLng' => [
                'display' => __('Around LatLng'),
                'type' => 'text',
                'instructions' => __('Force to first search around a specific latitude longitude. The option value must be provided as a string: `latitude,longitude` like `12.232,23.1`.'),
                'if' => [
                    'locationSearch' => 'equals lat_lng',
                ]
            ],
            'aroundRadius' => [
                'type' => 'text',
                'text_type' => 'number',
                'display' => __('Around radius'),
                'instructions' => __('Radius in meters to search around the latitude longitude. Otherwise a default radius is automatically computed given the area density.'),
                'if' => [
                    'locationSearch' => 'contains_any lat_lng_ip, lat_lng',
                ]
            ],
            'useDeviceLocation' => [
                'display' => __('Use device location'),
                'instructions' => __('Ask and use the device location.'),
                'type' => 'toggle',
                'default' => false,
                'width' => 50,
                'if' => [
                    'locationSearch' => 'contains_any lat_lng_ip, lat_lng',
                ]
            ],
            'insideBoundingBox' => [
                'type' => 'text',
                'display' => __('Inside bounding box'),
                'instructions' => __('Filters the results inside the area defined by the two extreme points of a rectangle. [See guide](https://www.algolia.com/doc/guides/managing-results/refine-results/geolocation/#filter-inside-rectanglepolygonal-area) or [API reference](https://www.algolia.com/doc/api-reference/api-parameters/insideBoundingBox/). <br><br>Format: `topRightLat, topRightLng, bottomLeftLat, bottomLeftLng`<br>Example: `"60, 16, 40, -4"`'),
                'if' => [
                    'locationSearch' => 'equals bounding_box',
                ]
            ],
            'insidePolygon' => [
                'type' => 'text',
                'display' => __('Inside polygon'),
                'instructions' => __('Filters the results inside the area defined by a shape. [See guide](https://www.algolia.com/doc/guides/managing-results/refine-results/geolocation/#filter-inside-rectanglepolygonal-area) or [API reference](https://www.algolia.com/doc/api-reference/api-parameters/insidePolygon/). <br><br>Format: `p1Lat, p1Lng, p2Lat, p2Lng, p3Lat, p3Lng...`'),
                'if' => [
                    'locationSearch' => 'equals polygon',
                ]
            ],
            'getRankingInfo' => [
                'type' => 'toggle',
                'display' => __('Get ranking info'),
                'instructions' => __('Controls whether the [_rankingInfo](https://www.algolia.com/doc/api-reference/api-methods/search/#method-response-_rankinginfo) object should be included in the hits.'),
                'default' => false,
                'width' => 50,
            ],
        ];
    }

    public function preload()
    {
        return [
            'placesAppId' => env('PLACES_APP_ID', false),
            'placesApiKey' => env('PLACES_API_KEY', false)
        ];
    }

    /**
     * The blank/default value
     *
     * @return array
     */
    public function defaultValue()
    {
        return null;
    }

    /**
     * Pre-process the data before it gets sent to the publish page
     *
     * @param mixed $data
     * @return array|mixed
     */
    public function preProcess($data)
    {
        return $data;
    }

    /**
     * Process the data before it gets saved
     *
     * @param mixed $data
     * @return array|mixed
     */
    public function process($data)
    {
        return $data;
    }
}


src/Fieldtypes/Places.php
<?php

namespace Spatie\StatamicAlgoliaPlaces\Fieldtypes;

use Statamic\Facades\Antlers;
use Statamic\Fields\Fieldtype;

class Places extends Fieldtype
{
    protected $icon = 'pin';

    protected function configFieldItems(): array
    {
        return [
            'placeholder' => [
                'type' => 'text',
                'display' => __('Placeholder'),
                'default' => __('Start typing your address'),
            ],
            'searchType' => [
                'type' => 'select',
                'display' => __('Type'),
                'instructions' => __('Restrict the search results to a specific type'),
                'options' => [
                    'all' => 'All',
                    'city' => 'City',
                    'country' => 'Country',
                    'address' => 'Address',
                    'busStop' => 'Bus Stop',
                    'trainStation' => 'Train Station',
                    'townhall' => 'Townhall',
                    'airport' => 'Airport',
                ],
                'default' => 'all',
            ],
            'postcodeSearch' => [
                'type' => 'toggle',
                'display' => __('Postcode Search'),
                'instructions' => __('Only search in the postcode field'),
            ],
            'language' => [
                'type' => 'text',
                'display' => __('Language'),
                'instructions' => __('Change the default language of the results. Leave empty to use the default: the locale of the current site. You can pass two letters language codes (ISO 639-1).'),
                'default' => '',
                'width' => 50,
            ],
            'countries' => [
                'type' => 'taggable',
                'display' => __('Countries'),
                'instructions' => __('Change the countries to search in. Two letters country codes (ISO 639-1).'),
                'width' => 50,
            ],
            'locationSearch' => [
                'type' => 'select',
                'options' => [
                    'no' => __('No'),
                    'lat_lng' => __('Around LatLng'),
                    'lat_lng_ip' => __('Around LatLng via IP'),
                    'bounding_box' => __('Inside a bounding box'),
                    'polygon' => __('Inside a polygon'),
                ],
                'default' => 'lat_lng_ip',
                'display' => __('Search around a location?'),
            ],
            'aroundLatLng' => [
                'display' => __('Around LatLng'),
                'type' => 'text',
                'instructions' => __('Force to first search around a specific latitude longitude. The option value must be provided as a string: `latitude,longitude` like `12.232,23.1`.'),
                'if' => [
                    'locationSearch' => 'equals lat_lng',
                ]
            ],
            'aroundRadius' => [
                'type' => 'text',
                'text_type' => 'number',
                'display' => __('Around radius'),
                'instructions' => __('Radius in meters to search around the latitude longitude. Otherwise a default radius is automatically computed given the area density.'),
                'if' => [
                    'locationSearch' => 'contains_any lat_lng_ip, lat_lng',
                ]
            ],
            'useDeviceLocation' => [
                'display' => __('Use device location'),
                'instructions' => __('Ask and use the device location.'),
                'type' => 'toggle',
                'default' => false,
                'width' => 50,
                'if' => [
                    'locationSearch' => 'contains_any lat_lng_ip, lat_lng',
                ]
            ],
            'insideBoundingBox' => [
                'type' => 'text',
                'display' => __('Inside bounding box'),
                'instructions' => __('Filters the results inside the area defined by the two extreme points of a rectangle. [See guide](https://www.algolia.com/doc/guides/managing-results/refine-results/geolocation/#filter-inside-rectanglepolygonal-area) or [API reference](https://www.algolia.com/doc/api-reference/api-parameters/insideBoundingBox/). <br><br>Format: `topRightLat, topRightLng, bottomLeftLat, bottomLeftLng`<br>Example: `"60, 16, 40, -4"`'),
                'if' => [
                    'locationSearch' => 'equals bounding_box',
                ]
            ],
            'insidePolygon' => [
                'type' => 'text',
                'display' => __('Inside polygon'),
                'instructions' => __('Filters the results inside the area defined by a shape. [See guide](https://www.algolia.com/doc/guides/managing-results/refine-results/geolocation/#filter-inside-rectanglepolygonal-area) or [API reference](https://www.algolia.com/doc/api-reference/api-parameters/insidePolygon/). <br><br>Format: `p1Lat, p1Lng, p2Lat, p2Lng, p3Lat, p3Lng...`'),
                'if' => [
                    'locationSearch' => 'equals polygon',
                ]
            ],
            'getRankingInfo' => [
                'type' => 'toggle',
                'display' => __('Get ranking info'),
                'instructions' => __('Controls whether the [_rankingInfo](https://www.algolia.com/doc/api-reference/api-methods/search/#method-response-_rankinginfo) object should be included in the hits.'),
                'default' => false,
                'width' => 50,
            ],
        ];
    }

    public function preload()
    {
        return [
            'placesAppId' => env('PLACES_APP_ID', false),
            'placesApiKey' => env('PLACES_API_KEY', false)
        ];
    }

    /**
     * The blank/default value
     *
     * @return array
     */
    public function defaultValue()
    {
        return null;
    }

    /**
     * Pre-process the data before it gets sent to the publish page
     *
     * @param mixed $data
     * @return array|mixed
     */
    public function preProcess($data)
    {
        return $data;
    }

    /**
     * Process the data before it gets saved
     *
     * @param mixed $data
     * @return array|mixed
     */
    public function process($data)
    {
        return $data;
    }

    public function config(string $key = null, $fallback = null)
    {
        return parent::config($key, $fallback); // TODO: Change the autogenerated stub
    }
}
